ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixParticipant.java

```diff
@@ -481,9 +486,16 @@ public boolean exitMaintenanceMode() {
         throw new IllegalStateException("Cluster is not in maintenance mode");
       }
       try {
-        helixAdmin.manuallyEnableMaintenanceMode(clusterName, false, maintenanceModeReason, null);
+        MaintenanceRecord record = getLastMaintenanceRecord();
+        if (record.operationType.equals(MaintenanceRecord.OPERATION_TYPE_ENTER) && record.reason.equals(
+            maintenanceModeReason)) {
+          helixAdmin.manuallyEnableMaintenanceMode(clusterName, false, maintenanceModeReason, null);
+          logger.info("Cluster {} exits maintenance mode: {}", clusterName, maintenanceModeReason);
+        } else {
+          logger.error("Maintenance record is not expected, operation type: {}, reason: {}", record.operationType,
+              record.reason);
+        }
         inMaintenanceMode = false;
```

Comment:
I'm not sure how useful this `inMaintenanceMode variable is if we will be checking in realtime if we're in MM or not anyways

Reply:
save one roundtrip to zookeeper to get the maintenance info.